### PR TITLE
DimItem vendor data

### PR DESCRIPTION
**Pull request by nev-r**
- swap out the `isVendorItem` bool for `.vendor` data, allowing a DimItem to be better aware of its source.
this will enable comparison of vendor items later
- gives vendor items more unique IDs so that they can be picked out by ID easier, including for wishlists
- moves ada bounty fix into vendoritem maker where i think it belongs
- document my findings about vendoritems as i go

### CODE REVIEW

**Comment by nev-r** [comment]
i don't want to break d1 and don't know how to work with its vendors, so i directly translated the `!owner` condition that originally defined whether isVendorItem would have been true

File: src/app/inventory/store/d2-item-factory.ts [hunk 1]
```diff
@@ -329,9 +332,17 @@ export function makeItem(
     itemDef.iconWatermarkShelved ||
     undefined;
 
-  const collectible =
-    itemDef.collectibleHash && mergedCollectibles && mergedCollectibles[itemDef.collectibleHash];
+  // collection stuff: establish a collectedness state, and hashes leading to the collectible and source
+  const { collectibleHash } = itemDef;
+  let collectibleState: DestinyCollectibleState | undefined;
+  if (collectibleHash) {
+    collectibleState = mergedCollectibles?.[collectibleHash]?.state;
```
Comment (bhollis) [inline src/app/inventory/store/d2-item-factory.ts:339]: I do sorta wonder if we should move this back out to the display layer...

File: src/app/inventory/store/d2-item-factory.ts [hunk 2]
```diff
@@ -385,12 +395,7 @@ export function makeItem(
       itemDef.nonTransferrable || item.transferStatus === TransferStatuses.NotTransferrable
     ),
     canPullFromPostmaster: !itemDef.doesPostmasterPullHaveSideEffects,
-    id:
-      item.itemHash === 3675595381 // Fix for ada-1 bounties ... https://github.com/Bungie-net/api/issues/1522
-        ? '1'
-        : item.itemHash === 171866827
-        ? '4'
-        : item.itemInstanceId || '0', // zero for non-instanced is legacy hack
+    id: item.itemInstanceId || '0', // zero for non-instanced is legacy hack
```
Comment (bhollis) [inline src/app/inventory/store/d2-item-factory.ts:398]: Hmm, maybe time to clean this out (separately)

File: src/app/vendors/vendor-item.ts [hunk 3]
```diff
@@ -15,9 +15,14 @@ import { DimItem } from '../inventory/item-types';
...
 
 /**
- * A displayable vendor item.
+ * Not actually always a vendor item.
+ * This represents an item inside a vendor or an item contained in a plugset.
  */
 export class VendorItem {
+  /**
+   * creates a VendorItem that's not actually at a vendor. it's part of a plug set.
+   * this serves something about the collections interface?
```
Comment (bhollis) [inline src/app/vendors/vendor-item.ts:24]: Yeah, some things aren't really collections, like emotes.

File: src/app/vendors/vendor-item.ts [hunk 4]
```diff
@@ -38,11 +43,18 @@ export class VendorItem {
...
+   * this can include "instanced" stats plugs etc which describe the specifics
+   * of that copy they're selling
+   */
   static forVendorSaleItem(
     defs: D2ManifestDefinitions,
     buckets: InventoryBuckets,
     vendorDef: DestinyVendorDefinition,
     saleItem: DestinyVendorSaleItemComponent,
+    // all DIM vendor calls are character-specific. any sale item should have an associated character.
+    characterId: string,
```
Comment (bhollis) [inline src/app/vendors/vendor-item.ts:57]: This'll be super helpful

File: src/app/vendors/vendor-item.ts [hunk 5]
```diff
@@ -63,10 +75,16 @@ export class VendorItem {
       vendorItemDef,
       saleItem,
       itemComponents,
-      mergedCollectibles
+      mergedCollectibles,
+      true, // why?
```
Comment (bhollis) [inline src/app/vendors/vendor-item.ts:79]: Makes sense that vendor items are purchase-able?

File: src/app/vendors/vendor-item.ts [hunk 6]
```diff
@@ -130,19 +150,34 @@ export class VendorItem {
...
       vendorItemDef ? vendorItemDef.quantity : 1,
       mergedCollectibles
     );
 
     if (this.item) {
       this.item.hidePercentage = true;
+
+      // override the DimItem.id for vendor items, so they are each unique enough to identify
+      // (otherwise they'd get their vendor index as an id, which is only unique per-vendor)
+      this.item.id = `${vendorHash}-${this.key}`;
```
Comment (bhollis) [inline src/app/vendors/vendor-item.ts:175]: Awesome

**Review by bhollis — APPROVED** [review]

File: src/app/vendors/vendor-item.ts [hunk 7]
```diff
@@ -130,19 +150,34 @@ export class VendorItem {
       this.previewVendorHash = inventoryItem.preview.previewVendorHash;
     }
 
+    // Fix for ada-1 bounties ... https://github.com/Bungie-net/api/issues/1522
+    // changes their sort to match the game
+    if (itemHash === 3675595381 || itemHash === 171866827) {
+      this.key = itemHash === 3675595381 ? 1 : 4;
+    }
```
Comment (delphiactual) [inline src/app/vendors/vendor-item.ts:157]: much cleaner than my implementation!

**Pull request merged**